### PR TITLE
General Brave Ads house-keeping

### PR DESCRIPTION
**Pull request by tmancey**
Resolves https://github.com/brave/brave-browser/issues/32969

## Submitter Checklist:

- [x] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

**Review by kylehickinson — APPROVED** [review]
iOS++

File: components/brave_ads/core/public/client/ads_client.h [hunk 1]
```diff
@@ -62,20 +62,20 @@ class ADS_EXPORT AdsClient {
...
-                                  const std::string& confirmation_type,
-                                  base::Time time) const = 0;
+  // Cache an ad event for the specified instance |id|, |ad_type|,
+  // |confirmation_type| and |time|.
+  virtual void CacheAdEventForInstanceId(const std::string& id,
+                                         const std::string& ad_type,
+                                         const std::string& confirmation_type,
+                                         base::Time time) const = 0;
 
   // Get ad event history for the specified |ad_type| and |confirmation_type|.
```
Comment (szilardszaloki) [inline components/brave_ads/core/public/client/ads_client.h:72]: cache

File: components/brave_ads/core/public/client/ads_client.h [hunk 2]
```diff
@@ -62,20 +62,20 @@ class ADS_EXPORT AdsClient {
...
-  virtual std::vector<base::Time> GetAdEventHistory(
+  virtual std::vector<base::Time> GetCachedAdEvents(
       const std::string& ad_type,
       const std::string& confirmation_type) const = 0;
 
   // Reset ad event history for the specified |id|.
```
Comment (szilardszaloki) [inline components/brave_ads/core/public/client/ads_client.h:77]: cache

File: components/brave_ads/core/internal/user/user_interaction/ad_events/ad_event_handler_util.h [hunk 3]
```diff
@@ -17,6 +19,49 @@ bool HasFiredAdEvent(const AdInfo& ad,
                      const AdEventList& ad_events,
                      const ConfirmationType& confirmation_type);
 
+template <typename T>
+bool WasAdServed(const AdInfo& ad,
+                 const AdEventList& ad_events,
+                 const T& event_type) {
+  CHECK(mojom::IsKnownEnumValue(event_type));
```
Comment (szilardszaloki) [inline components/brave_ads/core/internal/user/user_interaction/ad_events/ad_event_handler_util.h:26]: You most probably don't need these `CHECK`s on `mojom::IsKnownEnumValue`, unless you explicitly turn off Mojom validation for these enums somehow (https://chromium.googlesource.com/chromium/src/+/HEAD/mojo/public/tools/bindings/README.md#enums).

**Review by szilardszaloki — APPROVED** [review]
LGTM

File: components/brave_ads/core/public/client/ads_client.h (hunk 1 above)
Comment (tmancey) [reply components/brave_ads/core/public/client/ads_client.h:72]: Fixed, thanks

File: components/brave_ads/core/public/client/ads_client.h (hunk 2 above)
Comment (tmancey) [reply components/brave_ads/core/public/client/ads_client.h:77]: Fixed, thanks

File: components/brave_ads/core/internal/user/user_interaction/ad_events/ad_event_handler_util.h (hunk 3 above)
Comment (tmancey) [reply components/brave_ads/core/internal/user/user_interaction/ad_events/ad_event_handler_util.h:26]: Thanks, will follow up in a separate issue/PR.

**Pull request merged**